pkg/volume/iscsi/iscsi.go

```diff
@@ -277,13 +305,74 @@ func portalMounter(portal string) string {
 	return portal
 }
 
-func getVolumeSource(spec *volume.Spec) (*v1.ISCSIVolumeSource, bool, error) {
+// get iSCSI volume info: readOnly and fstype
+func getISCSIVolumeInfo(spec *volume.Spec) (bool, string, error) {
 	if spec.Volume != nil && spec.Volume.ISCSI != nil {
-		return spec.Volume.ISCSI, spec.Volume.ISCSI.ReadOnly, nil
+		return spec.Volume.ISCSI.ReadOnly, spec.Volume.ISCSI.FSType, nil
 	} else if spec.PersistentVolume != nil &&
 		spec.PersistentVolume.Spec.ISCSI != nil {
-		return spec.PersistentVolume.Spec.ISCSI, spec.ReadOnly, nil
+		return spec.ReadOnly, spec.PersistentVolume.Spec.ISCSI.FSType, nil
+	}
+
+	return false, "", fmt.Errorf("Spec does not reference an ISCSI volume type")
+}
+
+// get iSCSI target info: target portal, portals, iqn, and lun
+func getISCSITargetInfo(spec *volume.Spec) (string, []string, string, int32, error) {
+	if spec.Volume != nil && spec.Volume.ISCSI != nil {
+		return spec.Volume.ISCSI.TargetPortal, spec.Volume.ISCSI.Portals, spec.Volume.ISCSI.IQN, spec.Volume.ISCSI.Lun, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.ISCSI != nil {
+		return spec.PersistentVolume.Spec.ISCSI.TargetPortal, spec.PersistentVolume.Spec.ISCSI.Portals, spec.PersistentVolume.Spec.ISCSI.IQN, spec.PersistentVolume.Spec.ISCSI.Lun, nil
+	}
+
+	return "", nil, "", 0, fmt.Errorf("Spec does not reference an ISCSI volume type")
+}
+
+// get iSCSI initiator info: iface and initiator name
+func getISCSIInitiatorInfo(spec *volume.Spec) (string, *string, error) {
+	if spec.Volume != nil && spec.Volume.ISCSI != nil {
+		return spec.Volume.ISCSI.ISCSIInterface, spec.Volume.ISCSI.InitiatorName, nil
+	} else if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.ISCSI != nil {
+		return spec.PersistentVolume.Spec.ISCSI.ISCSIInterface, spec.PersistentVolume.Spec.ISCSI.InitiatorName, nil
+	}
+
+	return "", nil, fmt.Errorf("Spec does not reference an ISCSI volume type")
+}
+
+// get iSCSI CHAP booleans
+func getISCSICHAPInfo(spec *volume.Spec) (bool, bool, error) {
```

Comment:
these accessors with multiple identical return types are a likely place for confusion (caller swaps DiscoveryCHAPAuth and SessionCHAPAuth, etc)

is there a reason not to use single field accessors? would be 10 accessor methods instead of 4, but much less error prone

Reply:
still outstanding